builtin/logical/aws/secret_access_keys.go

```diff
@@ -47,42 +46,41 @@ func secretAccessKeys(b *backend) *framework.Secret {
 	}
 }
 
-func genUsername(displayName, policyName, userType, usernameTemplate string) (ret string, warning string, err error) {
+func genUsername(displayName, policyName, userType, usernameTemplate string) (ret string, err error) {
 	switch userType {
-	case "iam_user":
-		// IAM users are capped at 64 chars; this leaves, after the beginning and
-		// end added below, 42 chars to play with.
+	case "iam_user", "assume_role":
+		// IAM users are capped at 64 chars
 		up, err := template.NewTemplate(template.Template(usernameTemplate))
 		if err != nil {
-			return "", "", fmt.Errorf("unable to initialize username template: %w", err)
+			return "", fmt.Errorf("unable to initialize username template: %w", err)
 		}
 
 		um := UsernameMetadata{
+			Type:        "IAM",
 			DisplayName: normalizeDisplayName(displayName),
 			PolicyName:  normalizeDisplayName(policyName),
 		}
 
 		ret, err = up.Generate(um)
 		if err != nil {
-			return "", "", fmt.Errorf("failed to generate username: %w", err)
+			return "", fmt.Errorf("failed to generate username: %w", err)
 		}
 		// To prevent template from exceeding IAM length limits
 		if len(ret) > 64 {
-			ret = ret[0:64]
-			warning = "the calling token display name/IAM policy name were truncated to 64 characters to fit within IAM username length limits"
+			return "", fmt.Errorf("the username generated by the template exceeds the IAM/STS username length limits of 64 chars")
 		}
 	case "sts":
-		// Capped at 32 chars, which leaves only a couple of characters to play
-		// with, so don't insert display name or policy name at all
 		up, err := template.NewTemplate(template.Template(usernameTemplate))
 		if err != nil {
-			return "", "", fmt.Errorf("unable to initialize username template: %w", err)
+			return "", fmt.Errorf("unable to initialize username template: %w", err)
 		}
 
-		um := UsernameMetadata{}
+		um := UsernameMetadata{
+			Type: "STS",
+		}
 		ret, err = up.Generate(um)
 		if err != nil {
-			return "", "", fmt.Errorf("failed to generate username: %w", err)
+			return "", fmt.Errorf("failed to generate username: %w", err)
 		}
```

Comment:
Can we add a 32 length check here (similar to the 64 length above) to guard against custom username templates that might generate longer than allowed usernames?

Reply:
Added!